src/plasmapy/particles/atomic.py

```diff
@@ -1225,17 +1193,76 @@ def stopping_power(
                 f"Please specify one of: total, electronic, or nuclear for component! (Got {component}.)"
             )
 
-        if energies is not None:
-            # Interpolate NIST data to the user-provided energy values. Uses log-log scale fed into a cubic spline.
-            cs = CubicSpline(
-                x=np.log(baseline_energies_data), y=np.log(relevant_stopping_data)
-            )
-
-            return energies, np.exp(
-                cs(np.log(energies.to("MeV").value))
-            ) * u.MeV * u.cm**2 / u.g
-
         return (
             baseline_energies_data * u.MeV,
             relevant_stopping_data * u.MeV * u.cm**2 / u.g,
         )
+
+
+def _stopping_power_interpolator(
+    incident_particle: Particle, material: str, component="total"
+) -> Callable[[u.Quantity], u.Quantity]:
+    r"""
+    Instantiate a SciPy interpolator that can be used to interpolate the stopping power of a given energy.
+    """
+
+    baseline_energies_data, relevant_stopping_data = _get_relevant_stopping_information(
+        incident_particle, material, component
+    )
+
+    cs = CubicSpline(
+        x=np.log(baseline_energies_data.value), y=np.log(relevant_stopping_data.value)
+    )
+
+    return lambda x: np.exp(cs(np.log(x.to(u.MeV).value))) * u.MeV * u.cm**2 / u.g
+
+
+@particle_input
+@validate_quantities(energies=u.MeV)
+def stopping_power(
+    incident_particle: Particle,
+    material: str,
+    energies: u.Quantity | None = None,
+    component="total",
+) -> tuple[u.Quantity, u.Quantity]:
+    """Calculate stopping powers for a provided particle in a provided material.
+
+    Parameters
+    ----------
+    incident_particle : |Particle|
+        The incident particle. Only protons and alpha particles are currently supported.
+
+    material : `str`
+        The material the particle is being stopped in. See notes for details on supported materials.
+
+    energies : `~astropy.units.Quantity`, default: See notes.
+        The particle kinetic energies for which the stopping power is calculated.
+
+    component : `str`, default: ``total``
+        The component of the stopping power to be calculated. Supported values are ``electronic``,
+        ``nuclear``, and ``total`` for the electronic, nuclear, and total energies, respectively.
+
+    Returns
+    -------
+    ``Tuple[u.Quantity, u.Quantity[u.MeV * u.cm**2 / u.g]]``
+        A two-tuple where the first element represents the energy values. The
+        second element is an array of the associated stopping powers.
+
+    Notes
+    -----
+    The data for stopping power is taken from the National Institute of
+    Standards and Technology's Stopping-Power and Range Tables :cite:p:`niststar:2005`.
+    Valid materials can be found on the NIST STAR website. The default energies
+    are taken from the data points in the STAR database.
+
+    """
+
+    if energies is None:
```

Comment:
I don't really understand the need to split this into several functions: can't you just: 

1) Get the relevant stopping power data
2) If energies is None, just return that
3) If energies is not None, then interpolate that over the requested energies and return that

It seems like it can be done in one function? Is this a function complexity issue?

Reply:
The `ParticleTracker` uses the  `_stopping_power_interpolator` function to get the interpolator for each material. These interpolators would otherwise be constructed on every time step-- I think this is similar to something we have in the grid object, something about a persistent interpolator?

Reply:
note to self: recombine into one function with `return_interpolator` keyword